### PR TITLE
Separate templates & some clean-up

### PR DESCRIPTION
**Pull request by janhak**
These are small changes, that mostly revolve around moving templating to another file, simplifying main method and clarifying options which might be passed to recipy cli (--verbose is not implemented). On their own, they probably improve things a little. However, the CLI requires some work and I am mostly opening this pull request to gather your opinions.

I am thinking of breaking it down as follows:
recipyCmd
├── **init**.py
├── recipycmd.py
├── commands
│   ├── **init**.py
│   ├── annotate.py
│   ├── gui.py
│   ├── latest.py
│   └── search.py
├── templates
│   ├── **init**.py
│   ├── annotate_tmpl.txt
│   ├── debug_tmpl.txt
│   ├── run_tmpl.text
│   └── templating.py
└── tests

The main recipycmd.py would only parse arguments and dispatch the calls to relevant functions. Sub-commands would have their own --help screens to clarify further which options apply to which. If this is the direction you think is sensible then I am happy to try and move it over. It should hopefully make it easier to maintain and expand. I plan to do it in test driven development style. I am strongly thinking of https://github.com/pallets/click as opposed to docopt, as they should play better with nested commands and dockerization, but if you would like to keep the current technology I don't mind. 

Another two bits of logic that have to find their place are functions: _change_date and latest_run they should both move perhaps to recipyCommon.utils as they are called from both gui and cli. What do you think?

### CODE REVIEW

**Comment by robintw** [comment]
This looks really nice - far _far_ better than the previous code!

I like the changes you've done so far, but I think the changes you've suggested would be even better - and I agree with moving to `click` (quite a few people have recommended that to me). If you're happy to go forward with a click-based CLI then that would be wonderful.

As for `_change_date` and `latest_run`, I agree, they should probably be moved to `recipyCommon.utils`.

So, basically, in summary: this is great, and please go ahead - I look forward to seeing the results. (Feel free to just continue pushing to this PR).

**Comment by JanChwiejczak** [comment]
Thanks! I will carry on and keep pushing. Love the encouragement, I am playing with tox and setting up automated testing against various versions of python, which is something I would love us to have. I will keep you updated.

**Comment by janhak** [comment]
I think I got click figured out finally. Please have a look here: 

https://github.com/janhak/recipycli

It is a stub for recipy command line tool, fully tested against multiple versions of python with CI running on Travis. While writing it I did spend quite a lot of time interacting with recipy cli in its current form and would tentatively offer two changes:
- `recipy annotate` rename to `recipy tag` typing annotate gets old very quickly. However, it is a beautiful word which perhaps communicates the idea better so happy with both - your call.
- `recipy search` is layered, difficult to figure out and slightly buggy. For example:
```
(anaconda3-4.1.0) ➜ recipy search newplot    
Run ID: bcd7cba8-a120-4153-815b-c5980406b898
Created by jan on 2016-09-19 09:43:19 UTC
Ran /home/jan/Code/Python/recipy_play/examples/example_script.py using /home/jan/.pyenv/versions/anaconda3-4.1.0/bin/python
Environment: Linux-4.4.0-34-generic-tuxonice-x86_64-with-debian-stretch-sid, python 3.5.1 |Anaconda custom (64-bit)| (default, Jun 15 2016, 15:32:45) 
Libraries: recipy v0.3.0, pandas v0.18.1
Exception: (OSError) File b'data.csv' does not exist
Inputs:
/home/jan/Code/Python/recipy_play/examples/data.csv (None)
Outputs:
none
```
Where exaclty is newplot here?  Obviously, I omitted the extension (*.pdf) - as an end user I woud be mighty confused. Next I went:
```
(anaconda3-4.1.0) ➜ recipy search -f newplot.pdf
No results found
```
not encouraging! Same goes for regex, difficult to use without understanding how fields are being matched. If I regex `new.*` I would expect to find a match with `newplot.pdf` in:
```
/home/jan/Code/Python/recipy_play/examples/newplot.pdf (071ad1f637b28b1d865d89a04b1c2700e5e41b29)
/home/jan/Code/Python/recipy_play/examples/output3.csv (ebfe8e74d0686f26a0582a8b28f4dc4b771f7d6f)
```
Sadly `recipy search -r 'new.*'` finds no results. I Would propose something among these lines:
```
(cli) ➜  recipy search --help
Usage: recipy search [OPTIONS] FILENAME

  Search for FILENAME among all inputs and outputs recorded by ReciPy. By
  default, it returns partial matches and most recent run. You can use
  regular expressions in FILENAME.

  For example: recipy search -a --outputs newplot
  Will look through outputs of all runs and try to find a file containing
  newplot in its name.

Options:
  -e, --exact     Match FILENAME exactly.
  -i, --id        Match based on a fragment of run id.
  -f, --filepath  Match all files from the same directory as FILENAME.
  -a, --all       Return all results.
  -j, --json      Return results as JSON.
  --inputs        Restrict search to run inputs only.
  --outputs       Restrict search to run outputs only.
  --help          Show this message and exit.
``` 
I have real problems wording these help pages well and would welcome input on them. But the search I had in mind just tries to find something as eagerly as possible. Thoughts, Ideas?

As a next step, I thought about moving this exploratory work to recipy branch and integrate it. I also peeked into the innards of ReciPy, some smart stuff going on there - love it!

**Comment by robintw** [comment]
That looks wonderful - I haven't really used click before, but I can understand the code nice and easily, and it looks very clean and well-separated. If you're happy to start putting together a PR then that'd be great.

A few responses to your specific questions:

1. I agree about some of the issues with the word `annotate` - but I like the idea that it conveys that you can put 'notes' about the run in there, whereas (in my mind, at least) `tag` gives the impression of putting a very short word or phrase. I can see the benefits both ways though: @jvdzwaan, do you have a view here? One alternative might be to allow both (is there an easy way in `click` to have aliases for commands?) - but that might be more confusing than helpful.

2. I agree that the search functionality is complicated and buggy (though you should have seen what it was like before v0.3!), and in general I like your proposed solutions. _However_ (and, unfortunately, it's a fairly big however...) this doesn't take into account anything to do with searching based on hashes of files. At the moment, the default behaviour is to search using the file hash, if you give recipy a filename that exists (recipy will calculate the hash of that file and then search for that hash in the database). To search by filename/path you have to use other arguments such as `--filepath`/`-p`. This is the intended behaviour at the moment anyway - there may well be bugs in there!

So, the question is: how should we best deal with searching based on hashes and searching based on filenames? They are quite different things, but both come under the general concept of `search`. I had wondered before about splitting them into two separate commands - but I couldn't think of what to call them. Generally I feel that hash-based searching is more robust (you will only ever find the run that actually created that file, with no false positives) - but we need to deal sensibly with situations where no runs are found - and situations where there are no hashes stored because of configuration options! It gets rather complicated rather quickly...

Any thoughts you have on that would be very welcome!

Also, just to make you aware, @mikej888 is working on developing automated testing for recipy (he's part of the Software Sustainability Institute who are working on testing for recipy, as I mentioned in my talk). He's done some work on testing the current command-line interface at https://github.com/mikej888/recipy/tree/recipy_test/ - so most (all?) of those tests should pass for a new CLI based on click (obviously excluding any bits of the interface that we deliberately change). Just wanted to let you both know so you don't duplicate work.

**Comment by jvdzwaan** [comment]
1. 'note' is almost as short as 'tag'. I agree tag sounds like keywords only. Another option would be to allow something like -n or -a instead of having to type (or maybe both options). 

2. do there really need to be all the different search options in the CLI? In the GUI, everything is always searched using regexes if I remember correctly. Can't we do something similar for the CLI? I may be missing something here, but that's what I think :) The hashes part sounds a little more complicated, but is it really necessary for a file to exist to search for its hash? (In the GUI it is easy, because everything is clickable. Nobody is going to type a complete hash, I get that, but a hash is just a string that can be searched for, also by typing part of the hash.)

**Comment by robintw** [comment]
Its interesting: the GUI and CLI have different approaches to searching, and we should possibly try to make them similar _in some ways_ (but making sure we play to the strengths of each interface).

The idea for the hash searching is that if you have something like `awesomegraph.png` then you can run `recipy search awesomegraph.png` and it will tell you what run created it, even if when it was created it was called `graph.png`, because it's searching by the hash value.

I'm not entirely sure how I'd go about doing this in the GUI. Would I need to calculate the hash of `awesomegraph.png` manually (how? maybe we need a `recipy hash` command to calculate a hash in the same way used by recipy internally) and then put that into the search bar in the GUI? Is there (or could there be?) an easier way of doing this?

(More comments coming when I'm feeling less ill - but please continue discussing amongst yourselves).

**Comment by janhak** [comment]
On testing luckily we don't overlap what @mikej888 is having in mind is high-level integration tests which are great, and as long as I keep the command interface the same they should return same values. However, if we re-define some functionality with search some test might have to change.

On annotate I am happy to keep the current name but note gets my vote.

Search is a real problematic one, I completely misunderstood the intention of it and currently that intention is probably not explicit enough. I thought that it should work similar to GUI, basically just try to find as much as possible and if someone wants to narrow it down use the -e --exact flag. But you illustrate the point clearly with files that have been renamed. I don't have an answer let me think slightly.

Thanks @robintw and @jvdzwaan great input!

**Comment by janhak** [comment]
I will carry on working on other commands and their tests and will come back to search last.

My thoughts on search are:
0. Really like how GUI handles it but it is a bit irrelevant to cli as people will not search using complete paths.
1. If we keep search by hash as the default behaviour it must be obvious to end user. The option to search by the name of the file should be then available with a flag. 
2. Change the default behaviour to search by filename. It should have an option --hash -h which will guarantee search by hash. Also, if user types `recipy search awesomegraph.png` and the search can't find any runs with that file but it exists in the directory, it should try to hash it and re-run the search. If it finds something it should return that run with the following message affixed: `Couldn't find awsomegraph.png in ReciPy runs. Found run with graph.png which has the same hash value. Perhaps you have renamed it?` Such command might look like:
```
(cli) ➜  recipycli_play recipy search --help
Usage: recipy search [OPTIONS] FILENAME

  Search for FILENAME among all inputs and outputs recorded by ReciPy. By
  default, it returns partial matches and most recent run. You can use
  regular expressions in FILENAME.

  For example: recipy search -a --outputs newplot
  Will look through outputs of all runs and try to find a file containing
  newplot in its name.

Options:
  -h, --hash  Match using hash value of FILENAME. Will find original run even
              if the name of a file has changed.
  -a, --all   Return all runs.
  -j, --json  Return results as JSON.
  --inputs    Restrict search to run inputs only.
  --outputs   Restrict search to run outputs only.
  --help      Show this message and exit.
```
3. In both cases we should probably drop search by id option, this is a confused concept, we have a command designed to find runs associated with some files (whether by hash or name) but we also use it to find runs by their id. This should be called `recipy log` and should be a broader version of `recipy latest`. `recipy log` called with no values returns the latest run, `recipy log -id 345bg` returns runs which id's start with 345bg, `recipy log --all` returns all runs. 

Thoughts?

**Comment by janhak** [comment]
After giving it yet more thought I think the best course of action is to keep search exactly as it is for time being as not to interfere with @mikej888 and revisit that outside of this pull request. Otherwise, I am worried I will become overwhelmed with the task! What is done so far, has merit on its own.

The plan is to move remaining functions, cleanup and test all the code as is and not to change any functionality. Perhaps the only thing I would like to know is your thought of changing the latest to log as outlined above?

Thanks for all your inputs.

File: recipyCmd/cmd_annotate.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+
+@click.command('annotate', short_help='Add a note to the latest run.')
+@click.option('--id', '-i', help='Tag run which id starts with TEXT.')
+@pass_config
+def cmd(config, id):
+    """Add a note to the latest run using default text editor.
+     You can also tag older runs using --id.
+
+    \b
```
Comment (robintw) [inline recipyCmd/cmd_annotate.py]: Is this \b intentional?

Comment (janhak) [inline recipyCmd/cmd_annotate.py]: Hope you are feeling better and thanks for lookign through the code. I should have the refactoring wrapped up by tonight and tested, I will keep the search as it is and only iron out the gremlins.

The \b here is to prevent the re-wrappping of the lines that follow. It will be deleted from the help message dispalyed. You can see the full explanation here http://click.pocoo.org/5/documentation/

**Comment by janhak** [comment]
This is finished for the time being. Tested against python version 2.6, 2.7, 3.4 and 3.5 and mostly covered with automatic tests. Search command is changed minimally:
- fuzzy and regex rolled into one option
- accepts flags for inputs and outputs (as per TODO)
Not entirely happy with the code in the search command, needs further breaking down. Colorama should be automatically activated in windows environment by click, the call for it is removed. Have a play see whether you can break it, if not it can be merged.

**Comment by robintw** [comment]
Thanks - this all sounds great.

Just wanted to let you know: I'm away for a long weekend, but will look at this when I'm back.

**Comment by janhak** [comment]
Hey Robin, 

Not sure how are your time commitments with recipy at the moment. I would be keen to finish the work at some point. However, I am very mindful that someone was working on automated testing last time we spoke and wary of changing any of the interfaces. How is that going?

Is the code clean up in this pull interesting or relevant, or should we just close it what are your view?

Thanks

Jan

**Comment by robintw** [comment]
Hi Jan,

I definitely am still interested in this, I just haven't had chance to give it the attention it deserves before merging. The work on the automated testing has been finished, but again I haven't had a proper chance to go through it and test this PR with the automated test code. I really do apologise about this - a combination of my health, work, and my wife's difficult pregnancy, has prevented me doing much open-source work.

I'm starting to get back into some open-source work - and was aiming to spend some time on this over the next week or so. I'll try really hard to make sure that happens!

**Comment by janhak** [comment]
Robin!

Don't worry mate, I am glad that you are getting your head above the water. But they are always priorities in life. I wish much health to you and your growing family. Other things pale in comparison, frankly. I will keep it open, but please don't feel pressurised to look, it will happen when it happens!
